src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs

```diff
@@ -180,8 +205,30 @@ private static int VerifyCertChain(IntPtr storeCtxPtr, IntPtr arg)
 
                         using (X509Certificate2 leafCert = new X509Certificate2(leafCertPtr))
                         {
-                            success = chain.Build(leafCert);
-                            AddChannelBindingToken(leafCert, arg);
+                            var serverCallback = easy._handler._serverCertificateValidationCallback;
+                            if (serverCallback == null)
+                            {
+                                SslPolicyErrors errors = CertificateValidation.BuildChainAndVerifyProperties(
+                                    chain, leafCert, checkCertName: false, hostName: null);
+                                success = errors == SslPolicyErrors.None;
```

Comment:
I'm fine with any of these approaches:

* Let libcurl do the validation, we only do it for a server callback
* Let libcurl do the validation when there's no callback, we do it when there is one.
* Tell libcurl to not bother, we do the validation quickly in the accelerated path, and save the answer for the callback path.
    * This wouldn't have any of the previous perf problems since we're not doing a bunch of native -> managed -> native object rehydrations.

Really the difference between the first two is "codify the behavior with a test".  If curl doing the validation always means a hostname match is blocked, we need to have them not do it when a user callback is presented, since the user could be forgiving the hostname violation.